infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,307 @@
+                                                        .build());
+                                    } else {
+                                        // Get total number of addresses needed to transfer.
+                                        long numAddressesToTransfer = unknownAddressesInRange.size();
+                                        // Create a batch stream.
+                                        Stream<Batch> batchStream = Lists
+                                                .partition(unknownAddressesInRange, batchSize)
+                                                .stream()
+                                                .map(groupedAddresses ->
+                                                        new Batch(groupedAddresses, Optional.empty()
+                                                        )
+                                                );
+                                        // Execute state transfer synchronously.
+                                        return synchronousStateTransfer(batchProcessor, batchStream)
+                                                .thenApply(result ->
+                                                        createStatusBasedOnTransferResult(result,
+                                                                numAddressesToTransfer));
+
+                                    }
+                                    // If a segment contains some other status -> return.
+                                } else {
+                                    return CompletableFuture.completedFuture(status);
+                                }
+                            });
+                    return CurrentTransferSegment
+                            .builder()
+                            .startAddress(segment.getStartAddress())
+                            .endAddress(segment.getEndAddress())
+                            .status(newStatus)
+                            .build();
+                }
+
+
+        ).collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Given a batch processor and a stream of batches, execute the state transfer.
+     *
+     * @param batchProcessor An instance of a batch processor.
+     * @param batchStream    A stream of batches.
+     * @return A completed future containing a result of total number of addresses transferred
+     * or an exception.
+     */
+    CompletableFuture<Result<Long, StreamProcessFailure>> synchronousStateTransfer(
+            StateTransferBatchProcessor batchProcessor, Stream<Batch> batchStream) {
+        Result<Long, StreamProcessFailure> accumulatedResult = Result.ok(0L);
+
+        Result<Long, StreamProcessFailure> resultOfTransfer =
+                batchStream.reduce(accumulatedResult, (resultSoFar, nextBatch) -> {
+                    BatchResult batchResult = CFUtils
+                            .getUninterruptibly(batchProcessor.transfer(nextBatch));
```

Comment:
since `getUninterruptibly` is a blocking call, please try to avoid using it. At least you can use `.join()`

Reply:
Done